### DISCUSSION TITLE
[Feature] Trash - sort by date deleted

### DISCUSSION THREAD

**Discussion by blublub** [immich-app/immich#10365] [Feature Request]
### I have searched the existing feature requests to make sure this is not a duplicate request.

- [X] Yes

### The feature

Hello
I would like to suggest a sort option for "trash" view so that the sorting is not by date of the image/assat but rather the date it was deleted - I think that might be helpful in some cases where a lot was deleted within a few days
thx

### Platform

- [ ] Server
- [X] Web
- [ ] Mobile

**Comment by Dheerajdot** [comment]
I believe it will add significant value to users' experience. I would love to work on this.

**Comment by interunivers** [comment]
Very much needed, for example when accidentally deleting a file when sorting duplicates and it deletes two pictures in a row due to slow internet.

**Reply by Youwenqwq** [reply]
Yes, I'm facig this problem too. Another useful feature is adding a “revoke” option when deleting a file.

**Comment by lscorcia** [comment]
Is there a workaround here via API or database query? I just deleted a picture accidentally while removing duplicates and I'm trying to find it back...

**Reply by lscorcia** [reply]
Ok, so you can do this:

`sudo docker exec -it immich_postgres psql --dbname=immich --username=postgres`

`select "assets"."id", "assets"."originalFileName", "assets"."deletedAt" from assets order by "assets"."deletedAt" limit 100;`

Then put the id at the end of the URL `https://<your_immich_server>/trash/photos/<id>`, visit it in your browser and you will be able to find it again. In the three dots menu then use the command "Restore".

Doesn't seem to work for RAW images though :/

**Reply by lscorcia** [reply]
Even better: you can use the UI! Just visit in the browser:
`https://<your_immich_server>/search?query={"trashedAfter":"2024-08-21T13:00:00"}`, replacing with the date-time of the approximate deletion. It will display the recently deleted items and you can restore them from there.

**Reply by caddish** [reply]
> Even better: you can use the UI! Just visit in the browser: `https://<your_immich_server>/search?query={"trashedAfter":"2024-08-21T13:00:00"}`, replacing with the date-time of the approximate deletion. It will display the recently deleted items and you can restore them from there.

Help !! I'm trying to use this to restore 11,000 files but there's no option to restore from multi select 😭

**Comment by redge76** [comment]
Same here. I deleted pictures by mistake and they were pretty hard to find in the trash as I did not know their date and I keep 365 days of trash. (did not know the query on trashedAfter trick)
Sorting by deletion date would make this kind of mistake easy to repair.

**Comment by Thyrfing89** [comment]
Just happend here! Would be very helpful!

**Comment by tobse91** [comment]
+1
Simple feature but it would be very handy.

**Comment by RenaKunisaki** [comment]
Seems like something that should be high priority. Right now if I accidentally delete something it's just gone, because there's no confirmation, and my trash is clogged with thousands of offline assets making it impossible to find the file.

**Reply by bo0tzz** [reply]
If this is high priority to you, feel free to make a PR.

**Reply by RenaKunisaki** [reply]
That's exactly the kind of condescending non-response that makes me wary of using Immich at all. If a long-standing UI issue that leads to data loss is *not* of high priority to maintainers, then what is? And this attitude tells me that any PR I do make is unlikely to be taken seriously.

It's not as if the issue is "add a function that does my laundry and pays my taxes". We're talking about what should be - for someone who's already familiar with the codebase - a fairly trivial change to prevent a fairly obvious and large usability problem.

**Reply by redge76** [reply]
Yes, I agree that's the kind of missing functionnality that make a feature pretty hard to use.  
Google photo sorts it's trash by deletion date.

edit: I previously thought it worked but no...

**Reply by bo0tzz** [reply]
> That's exactly the kind of condescending non-response that makes me wary of using Immich at all. 

It's not intended to be condescending, the maintainer team's priorities are simply elsewhere. Ask 10 users what should be high priority and you'll get 15 answers, the best tool we have there is to encourage people to contribute to the project to solve problems they're having

> If a long-standing UI issue that leads to data loss is not of high priority to maintainers, then what is? 

Data loss is absolutely a high priority, of course. I don't see how being unable to sort the trash would lead to data loss though. You also mentioned having many offline assets in there, which is understandably annoying and is something we obviously intend to change, but I don't see what that has to do with the priority of this trash sorting feature (other than as a workaround).

> And this attitude tells me that any PR I do make is unlikely to be taken seriously.

Why would a PR for this not be taken seriously, especially if I'm explicitly inviting you to? Sorting options are an obvious feature and PRs would be quite welcome.

> what should be - for someone who's already familiar with the codebase - a fairly trivial change 

Whether a change is trivial is rarely assessed accurately by someone who's not familiar with the codebase, and besides, arguments like that would quickly lead to dozens of demands for "trivial" changes taking away from other priorities.

**Reply by RenaKunisaki** [reply]
I'm just not sure how to interpret your initial reply as anything other than a snarky "well fix it yourself then".

The data loss aspect comes from the UI making it very easy to accidentally delete things (potentially without even noticing) and then very difficult to restore them. I'll admit it's probably not as big of an issue for most users, who probably don't have thousands of offline assets hanging out in there.

**Comment by Fgdou** [comment]
I guess it's also related to https://github.com/immich-app/immich/discussions/8377, not just for search but also for the trash

**Comment by bradyjoh** [comment]
Coming across this and would love an undo button after deleting, or sorting by date deleted in the trash. This is coming from not having configurable keybindings, so photos get trashed by accident sometimes and then I need to spend a couple minutes going through the trash to restore the one I put there on accident.

**Comment by magakis** [comment]
This feature is mandatory for the trash to actually make sense. You might be cleaning your library and accidentally delete a photo. If you lose the timeline that this photo was taken, then you can consider having lost this photo in the trash forever as well.

**Comment by timonrieger** [comment]
This one implements the feature request apparently: https://github.com/immich-app/immich/pull/22690

**Comment by smeredith** [comment]
I would have liked to have this feature today after accidentally deleting a photo without noticing what it looked like.

**Comment by Lomacar** [comment]
This is a pretty basic expectation of how a trash should work. I can only assume it hasn't been implemented from the beginning because there is some complication with how the trash works and/or it is not currently easy to record a deletion event in the database connecting the file and the media?

With AI coding these days it is so easy to contribute, so I am tempted, but there could be multiple ways to implement it and established developers would have the best opinions of how it should be done. It would be nice to have developers here give their opinions on the matter. Once there is some discussion you can practically throw that discussion into Cursor and it will write the code automatically.

**Comment by strobeflash** [comment]
The other day I accidentally deleted most of my assorted assets by shift selecting the wrong way. Clearly a layer 8 (user) error, not of the system. All assets were still available in the trash, as expected.

I couldn't find a restore procedure for accidental deletions like these and I aggree that it would be great if the trash could be sorted by deletion date. In the mean time you can find the steps I used to restore below, I hope this may be useful to another tortured soul.

[immich-restore.md](https://github.com/user-attachments/files/29216654/immich-restore.md)

**Comment by redge76** [comment]
Here are 2 other options

---
## Option 1 — Use the UI (for a small number of assets)
If only a few images were deleted, you can restore them manually using a search query.
### Search URL
https://demo.immich.app/search?query={"trashedAfter":"2026-06-23T09:00:00Z","trashedBefore":"2026-06-24T15:00:00Z"}
### Steps
1. Open the link in your browser
2. Browse the listed images
3. Open an image
4. Click the three-dot menu
5. Click **Restore**
### Pros / Cons
- ✅ Very simple  
- ✅ No API needed  
- ❌ Not practical for large amounts (e.g. 500+ images)
---

## Option 2 — Use the API (recommended for bulk restore)

### Step 1 — Search for trashed assets

Go to:
https://api.immich.app/endpoints/search/searchAssets
Use the following query:
    {"trashedAfter":"2026-06-23T09:00:00Z","trashedBefore":"2026-06-24T15:00:00Z"}
This returns a JSON response containing all matching deleted assets.


### Step 2 — Extract asset IDs

From the API response, extract all `id` values from:
    assets.items[]
Example:
    {"assets":{"items":[{"id":"6418c37d-35b0-4011-882d-36946bc00eb7"},{"id":"11b84a96-f89e-4059-988f-67cb50a807dd"}]}}

### Step 3 — Build the restore payload

    {"ids":["6418c37d-35b0-4011-882d-36946bc00eb7","11b84a96-f89e-4059-988f-67cb50a807dd"]}

### Step 4 — Restore assets

Send the payload to:
https://api.immich.app/endpoints/trash/restoreAssets

## Tip — Automate with an LLM

Prompt:
From this JSON, extract all asset IDs and generate a payload for the restoreAssets API

**Comment by distante** [comment]
I really hope this feature request gets pickup up
